### PR TITLE
Refactoring of AvailableModeUtils: extracted method, simplified code …

### PR DESCRIPTION
**Pull request by carloscaldas**
Improvements:
- Converted class to object
- Extracted method to avoid code duplication
- Code simplification: avoided redundant tests
- Tightened methods visibility level
- Removed unused method

---
This change is [Reviewable](https://reviewable.io/reviews/lbnl-ucb-sti/beam/1939)

### CODE REVIEW

File: src/main/scala/beam/utils/plan/sampling/AvailableModeUtils.scala [hunk 1]
```diff
@@ -50,81 +51,89 @@ object AvailableModeUtils extends LazyLogging {
...
+    Option(gerPersonCustomAttributes(person)).getOrElse {
+      val attribs: AttributesOfIndividual =
+        PopulationAdjustment.createAttributesOfIndividual(beamScenario, population, person, household)
       person.getCustomAttributes.put(PopulationAdjustment.BEAM_ATTRIBUTES, attribs)
       attribs
     }
-    setModesForPerson(person, population, permissibleModes, attributesOfIndividual)
+  }
+
+  private def gerPersonCustomAttributes(person: Person): AttributesOfIndividual = {
```
Comment (JustinPihony) [inline src/main/scala/beam/utils/plan/sampling/AvailableModeUtils.scala]: Typo: ger -> get

File: src/main/scala/beam/utils/plan/PlansBuilder.scala [hunk 2]
```diff
@@ -43,8 +43,7 @@ object PlansBuilder {
   val newHH: HouseholdsImpl = new HouseholdsImpl()
   val newHHAttributes: ObjectAttributes = newHH.getHouseholdAttributes
 
-  val modeAllocator: AvailableModeUtils.AllowAllModes =
-    new AvailableModeUtils.AllowAllModes
+  val modeAllocator: AvailableModeUtils.AllowAllModes.type = AvailableModeUtils.AllowAllModes
```
Comment (JustinPihony) [inline src/main/scala/beam/utils/plan/PlansBuilder.scala]: If this is truly a static class (functions don't rely on internal state), which it appears is the case, then the standard is to reference it directly. This technically works, but should it be in a variable or just imported?

    import AvailableModeUtils.AllowAllModes

then it can just be referenced:

    AllowAllModes.METHOD

File: src/main/scala/beam/utils/plan/sampling/AvailableModeUtils.scala [hunk 3]
```diff
@@ -50,81 +51,89 @@ object AvailableModeUtils extends LazyLogging {
...
     }
   }
 
+  private def getAvailableModesOfPerson(
+    person: Person,
+    population: Population,
+    permissibleModes: Seq[String]
+  ): Seq[String] = {
+    val excludedModes = getExcludedModesForPerson(population, person.getId.toString)
+    permissibleModes.filterNot(am => excludedModes.exists(em => em.equalsIgnoreCase(am)))
```
Comment (JustinPihony) [inline src/main/scala/beam/utils/plan/sampling/AvailableModeUtils.scala]: You lost the existence check on `excludedModes`. I assume this was a micro-optimization, but it is more optimized. It should probably stay

File: src/main/scala/beam/utils/plan/PlansBuilder.scala (hunk 2 above)
Comment (carloscaldas) [reply src/main/scala/beam/utils/plan/PlansBuilder.scala]: Actually this is not a method. is an `Object` inside another `Object`. that is why the word `type` is needed

File: src/main/scala/beam/utils/plan/sampling/AvailableModeUtils.scala (hunk 3 above)
Comment (carloscaldas) [reply src/main/scala/beam/utils/plan/sampling/AvailableModeUtils.scala]: In this case outcome is exactly the same. Benefits: smaller code with at same level of readability(IMO is better readability) and micro optimization.

File: src/main/scala/beam/utils/plan/sampling/AvailableModeUtils.scala (hunk 1 above)
Comment (carloscaldas) [reply src/main/scala/beam/utils/plan/sampling/AvailableModeUtils.scala]: fixed

File: src/main/scala/beam/utils/plan/PlansBuilder.scala (hunk 2 above)
Comment (carloscaldas) [reply src/main/scala/beam/utils/plan/PlansBuilder.scala]: fixed

File: src/main/scala/beam/utils/plan/PlansBuilder.scala [hunk 4]
```diff
@@ -123,7 +122,7 @@ object PlansBuilder {
     val permissibleModes = modeAllocator
       .getPermissibleModes(person.getSelectedPlan)
       .asScala
-    AvailableModeUtils.setAvailableModesForPerson(person, newPop, permissibleModes.toSeq)
+    setAvailableModesForPerson(person, newPop, permissibleModes.toSeq)
```
Comment (JustinPihony) [inline src/main/scala/beam/utils/plan/PlansBuilder.scala]: Actually, this one I'm not a fan of. I prefer taking it to the final object and then leaving it there when it comes to statics. Otherwise it makes it just a bit harder to trace where the method is defined - especially for new devs

Comment (carloscaldas) [reply src/main/scala/beam/utils/plan/PlansBuilder.scala]: This PR is a pre-step to fulfil the the issue https://github.com/LBNL-UCB-STI/beam/issues/1853

Having an "static" object to apply mutations to an object that should be immutable in the first place looks like spreading the problem around the system.

The final goal here is moving the responsibility yo the loader (`BeamScenarioLoader`). There is no reason to load scenario partially and in other steps fix the inconsistent scenario. 

To sum up: ScenarioLoader will be responsible to load persons (like is already is) and its attributes (like is partially is). Ideally the scenario will become immutable and once is loaded cannot change anymore

Comment (carloscaldas) [reply src/main/scala/beam/utils/plan/PlansBuilder.scala]: in other words: mutating a state passed as parameter is a bad practice.
```
setAvailableModesForPerson(person, newPop, permissibleModes.toSeq)
```
will be replaced with
```
person.availableModes = calculateAvailableModes(newPop, permissibleModes.toSeq)
```
when building the scenario

Comment (carloscaldas) [reply src/main/scala/beam/utils/plan/PlansBuilder.scala]: done

**Comment by carloscaldas** [comment]
test!
